src/libinjector/linux/linux_injector.c

```diff
@@ -0,0 +1,225 @@
+#include "linux_utils.h"
+#include "linux_debug.h"
+#include "methods/linux_shellcode.h"
+
+bool check_userspace_int3_trap(injector_t injector, drakvuf_trap_info_t* info)
+{
+    // check CPL
+    unsigned long int CPL = (info->regs->cs_sel & 3);
+    PRINT_DEBUG("CPL 0x%lx\n", CPL);
+
+    if ( CPL != 0)
+    {
+        PRINT_DEBUG("Inside INT3 userspace\n");
+    }
+    else
+    {
+        PRINT_DEBUG("INT3 received but CPL is not 0x3\n");
+        return false;
+    }
+
+    if ( info->proc_data.pid != injector->target_pid )
+    {
+        PRINT_DEBUG("INT3 received but '%s' PID (%u) doesn't match target process (%u)\n",
+            info->proc_data.name, info->proc_data.pid, injector->target_pid);
+        return false;
+    }
+
+    if (info->regs->rip != info->trap->breakpoint.addr)
+    {
+        PRINT_DEBUG("INT3 received but BP_ADDR (%lx) doesn't match RIP (%lx)",
+            info->trap->breakpoint.addr, info->regs->rip);
+        assert(false);
+    }
+
+    if (injector->target_tid && (uint32_t)info->proc_data.tid != injector->target_tid)
+    {
+        PRINT_DEBUG("INT3 received but '%s' TID (%u) doesn't match target process (%u)\n",
+            info->proc_data.name, info->proc_data.tid, injector->target_tid);
```

Comment:
`drakvuf_check_return_context` ?

Reply:
this seems to made for check to function calls in mind ( it is checking `rsp` also ), do you want me to change that function definition and call that here?